File: geeksforgeeks/Longest valid Parentheses.c
#include <stdio.h>
#include <string.h>
#define MAX_N 100000

char string[MAX_N + 2];
int stack[MAX_N], closed[MAX_N];

int longest_valid_parentheses(int n, char string[], int stack[], int closed[]){
    int i, j, length = 0, max_length = 0, stack_size = 0;
    for(i = 0; i < n; i++){
        closed[i] = 0;
        if(string[i] == '('){
            stack[stack_size++] = i;
        }else if(stack_size > 0){
            j = stack[--stack_size];
            closed[i] = closed[j] = 1;
        }
    }
    for(i = 0; i < n; i++){
        if(closed[i]){
            length += 1;
            if(length > max_length)
                max_length = length;
        }else{
            length = 0;
        }
    }
    return max_length;
}

int main(){    
    int t, n;
    scanf("%d", &t);
    while(t-- > 0){
        scanf("%s", string);
        n = (int) strlen(string);
        printf("%d\n", longest_valid_parentheses(n, string, stack, closed));
    }
    return 0;
}

File: geeksforgeeks/Longest valid Parentheses.py
def longest_valid_parentheses(string):
    n = len(string)
    closed, stack, length, max_length = [False] * n, [], 0, 0
    for i in range(n):
        if string[i] == '(':
            stack.append(i)
        elif stack:
            j = stack.pop()
            closed[i], closed[j] = True, True
    for i in range(n):
        if closed[i]:
            length += 1
            max_length = length if length > max_length else max_length
        else:
            length = 0
    return max_length

if __name__ == '__main__':
    for i in range(int(input())):
        print(longest_valid_parentheses(input().strip()))
